Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
+			//carrier docked somewhere, undock into that base. POBs require special handling.
+			CUSTOM_BASE_IS_DOCKED_STRUCT POBcheckStruct;
+			POBcheckStruct.iClientID = carrierClientID;
+			POBcheckStruct.iDockedBaseID = 0;
+			Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_BASE_IS_DOCKED, &POBcheckStruct);
+			if (POBcheckStruct.iDockedBaseID)
+			{
+				CUSTOM_BASE_BEAM_STRUCT POBbeamStruct;
+				POBbeamStruct.iClientID = client;
+				POBbeamStruct.iTargetBaseID = POBcheckStruct.iDockedBaseID;
+				Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_BASE_BEAM, &POBbeamStruct);
+			}
+			else
+			{
+				uint baseID;
+				pub::Player::GetBase(carrierClientID, baseID);
+				pub::Player::ForceLand(client, baseID);
+			}
 		}
-
-		mobiledockClients.erase(client);
-	}
-
-	// Land ship to proxy base in carrier's system if they are in different systems.
-	if (mobiledockClients[client].carrierSystem == -1)
-	{
-		if (!mobiledockClients[client].carrierDied)
+		else
 		{
-			string scProxyBase = HkGetPlayerSystemS(HkGetClientIdFromCharname(mobiledockClients[client].wscDockedWithCharname)) + "_proxy_base";
-			if (pub::GetBaseID(mobiledockClients[client].proxyBaseID, scProxyBase.c_str()) == -4)
+			//check if carrier is currently mid-transit, if so, force dock him back to proxy base and abort;
+			CUSTOM_IN_WARP_CHECK_STRUCT warpCheck;
+			warpCheck.clientId = carrierClientID;
+			Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_IN_WARP_CHECK, &warpCheck);
+			if (warpCheck.inWarp)
 			{
-				PrintUserCmdText(client, L"No proxy base in system detected. Contact a developer about this please. Required base: %s", scProxyBase);
+				pub::Player::ForceLand(client, Players[client].iLastBaseID);
+				PrintUserCmdText(client, L"ERR Can't undock while carrier is in hyperspace transit.");
+				returncode = SKIPPLUGINS_NOFUNCTIONCALL;
 				return;
 			}
-		}
 
-		pub::Player::ForceLand(client, mobiledockClients[client].proxyBaseID);
+			//teleport the player using same method as jumpdrives
+			CUSTOM_JUMP_CALLOUT_STRUCT jumpData;
+			Vector pos;
+			Matrix ori;
+			uint iSystemID;
 
-		// Send the message because if carrier goes to another system, docked ships remain in previous with outdated system navmap. We notify client about it is being updated.
-		PrintUserCmdText(client, L"Navmap updated successfully.");
+			pub::SpaceObj::GetLocation(carrierShipID, pos, ori);
+			pub::SpaceObj::GetSystem(carrierShipID, iSystemID);
 
-		// Update current system stat in player list to be displayed relevantly.
-		Server.BaseEnter(mobiledockClients[client].proxyBaseID, client);
-		Server.BaseExit(mobiledockClients[client].proxyBaseID, client);
-		wstring wscCharFileName;
-		HkGetCharFileName((const wchar_t*)Players.GetActiveCharacterName(client), wscCharFileName);
-		wscCharFileName += L".fl";
-		CHARACTER_ID cID;
-		strcpy(cID.szCharFilename, wstos(wscCharFileName.substr(0, 14)).c_str());
-		Server.CharacterSelect(cID, client);
+			pos.x += ori.data[0][1] * set_iMobileDockOffset;

```

Comment:
Unless I am mistaken, I don't think that's how Matrixes work. What is this doing?

Reply:
Spawns the ship `set_iMobileDockOffset` metres above the ship.